lib/providers/portion/PortionProvider.ts

```diff
@@ -0,0 +1,255 @@
+      resolvedTokenIn && resolvedTokenOut
+        ? await this.portionFetcher.getPortion(
+          sharedInfo.tokenInChainId,
+          resolvedTokenIn.wrapped.address,
+          sharedInfo.tokenOutChainId,
+          resolvedTokenOut.wrapped.address
+        )
+        : undefined;
+
+    return getPortionResponse ?? GET_NO_PORTION_RESPONSE;
+  }
+
+  async getPortionForTokenOut(sharedInfo: QuoteRequestInfo, tokenOutAmount: string): Promise<FullPortionPayload> {
+    // we will need to call token fetcher to resolve the tokenIn and tokenOut
+    // there's no guarantee that the tokenIn and tokenOut are in the token address
+    // also the tokenIn and tokenOut can be native token
+    // portion service only accepts wrapped token address
+    const [_, resolvedTokenOut] = await this.resolveTokenInAndTokenOut(sharedInfo);
+
+    const getPortionResponse = await this.getPortion(sharedInfo);
+
+    const tokenOutCurrencyAmount = resolvedTokenOut
+      ? CurrencyAmount.fromRawAmount(resolvedTokenOut, tokenOutAmount)
+      : undefined;
+    const portionAmount = this.getPortionAmount(tokenOutCurrencyAmount, getPortionResponse, resolvedTokenOut);
+    const portionAdjustedTokenOutAmount = this.getPortionAdjustedTokenOutAmount(
+      sharedInfo.type,
+      tokenOutCurrencyAmount,
+      portionAmount
+    );
+    return {
+      portionResponse: getPortionResponse,
+      portionAmount: portionAmount,
+      portionAdjustedTokenOutAmount: portionAdjustedTokenOutAmount,
+    } as FullPortionPayload;
+  }
+
+  async getPortionAdjustedQuote(
+    sharedInfo: QuoteRequestInfo,
+    quote: string,
+    quoteGasAdjusted: string,
+    tokenOutAmount: string
+  ): Promise<FullPortionPayload> {
+    // we will need to call token fetcher to resolve the tokenIn and tokenOut
+    // there's no guarantee that the tokenIn and tokenOut are in the token address
+    // also the tokenIn and tokenOut can be native token
+    // portion service only accepts wrapped token address
+    const [resolvedTokenIn, resolvedTokenOut] = await this.resolveTokenInAndTokenOut(sharedInfo);
+
+    // portion service has no concept of native currency, need to pass in wrapped
+    const getPortionResponse =
+      resolvedTokenIn && resolvedTokenOut
+        ? await this.portionFetcher.getPortion(
+            sharedInfo.tokenInChainId,
+            resolvedTokenIn.wrapped.address,
+            sharedInfo.tokenOutChainId,
+            resolvedTokenOut.wrapped.address
+          )
+        : undefined;
+
+    const quoteCurrency = sharedInfo.type === TradeType.EXACT_INPUT ? resolvedTokenOut : resolvedTokenIn;
+    const quoteCurrencyAmount = quoteCurrency ? CurrencyAmount.fromRawAmount(quoteCurrency, quote) : undefined;
+    const quoteGasAdjustedCurrencyAmount = quoteCurrency
+      ? CurrencyAmount.fromRawAmount(quoteCurrency, quoteGasAdjusted)
+      : undefined;
+
+    const requestCurrency = sharedInfo.type === TradeType.EXACT_INPUT ? resolvedTokenIn : resolvedTokenOut;
```

Comment:
we usually will call the non quote token `amountToken` or when it's a currency, `amount`